src/coreclr/src/vm/reflectioninvocation.cpp

```diff
@@ -2174,6 +1997,269 @@ lExit: ;
+ * pfCtorIsPublic - [required, null-init] whether the parameterless ctor is public
+ * ==========
+ * This method will not run the type's static cctor.
+ * This method will not allocate an instance of the target type.
+ */
+void QCALLTYPE RuntimeTypeHandle::GetActivationInfo(
+    QCall::ObjectHandleOnStack pRuntimeType,
+    PCODE* ppfnAllocator,
+    void** pvAllocatorFirstArg,
+    PCODE* ppfnCtor,
+    BOOL* pfCtorIsPublic
+)
+{
+    CONTRACTL{
+        QCALL_CHECK;
+        PRECONDITION(CheckPointer(ppfnAllocator));
+        PRECONDITION(CheckPointer(pvAllocatorFirstArg));
+        PRECONDITION(CheckPointer(ppfnCtor));
+        PRECONDITION(CheckPointer(pfCtorIsPublic));
+        PRECONDITION(*ppfnAllocator == NULL);
+        PRECONDITION(*pvAllocatorFirstArg == NULL);
+        PRECONDITION(*ppfnCtor == NULL);
+        PRECONDITION(*pfCtorIsPublic == NULL);
+    }
+    CONTRACTL_END;
+
+    TypeHandle typeHandle = NULL;
+
+    // If we're activating __ComObject.
+    void* pClassFactory = NULL;
+
+    BEGIN_QCALL;
+
+    {
+        GCX_COOP();
+
+        // We need to take the RuntimeType itself rather than the RuntimeTypeHandle,
+        // as the COM CLSID is stored in the RuntimeType object's sync block, and we
+        // might need to pull it out later in this method.
+        typeHandle = ((REFLECTCLASSBASEREF)pRuntimeType.Get())->GetType();
+    }
+
+    ValidateTypeAbleToBeInstantiated(typeHandle, false /* fGetUninitializedObject */);
+
+    MethodTable* pMT = typeHandle.AsMethodTable();
+    PREFIX_ASSUME(pMT != NULL);
+
+#ifdef FEATURE_COMINTEROP
+    // COM allocation can involve the __ComObject base type (with attached CLSID) or a
+    // VM-implemented [ComImport] class. For GetUninitializedObject, we block all COM types.
+    // For CreateInstance, the flowchart is:
+    //   - For __ComObject,
+    //     .. on Windows, bypass normal newobj logic and use ComClassFactory::CreateInstance.
+    //     .. on non-Windows, treat as a normal class, type has no special handling in VM.
+    //   - For [ComImport] class, treat as a normal class. VM will replace default
+    //     ctor with COM activation logic on supported platforms, else ctor itself will PNSE.
+    if (pMT->IsComObjectType())
+    {
+        if (IsComObjectClass(typeHandle))
+        {
+#ifdef FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+            // Need to enter cooperative mode to manipulate OBJECTREFs
+            GCX_COOP();
+            SyncBlock* pSyncBlock = pRuntimeType.Get()->GetSyncBlock();
+            pClassFactory = (void*)pSyncBlock->GetInteropInfo()->GetComClassFactory();
+#endif // FEATURE_COMINTEROP_UNMANAGED_ACTIVATION
+
+            if (pClassFactory == NULL)
+            {
+                COMPlusThrow(kInvalidComObjectException, IDS_EE_NO_BACKING_CLASS_FACTORY);
+            }
+        }
+    }
+#endif // FEATURE_COMINTEROP
+
+    pMT->EnsureInstanceActive();
+
+    // All checks passed! Pass parameters back to the caller.
```

Comment:
This is almost true. There is still `Arg_NoDefCTorWithoutTypeName` failure path below.

I think you can:
- Move `pMT->EnsureInstanceActive();` to the very end of the method once everything is really done
- Move the special COM interop handling together inside one ifdef

ie:
```
#ifdef FEATURE_COMINTEROP
    if (IsComObjectClass(typeHandle))
    {
...
    }
    else
#endif
    if (pMT->IsNullable())
    {
     ....
    }
    else
    {
    ...
    }

    pMT->EnsureInstanceActive();

    END_QCALL;
```
```

Reply:
Are there potential negative consequences to calling `getNewHelperStatic` or `GetDefaultConstructor->GetMultiCallableAddrOfCode` _before_ the call to `EnsureInstanceActive`?

Reply:
There should not be. `EnsureInstanceActive` has to run before the relevant managed code runs. It does not need to run to prepare the managed code.